[MediaSource] Use an AVSampleBufferVideoRenderer when a video receiver endpoint is available

aestes opened the pull request:
#### d7df52b60ed83af41b0534cca154a3ac5a0533de
<pre>
[MediaSource] Use an AVSampleBufferVideoRenderer when a video receiver endpoint is available
<a href="https://bugs.webkit.org/show_bug.cgi?id=269621">https://bugs.webkit.org/show_bug.cgi?id=269621</a>
<a href="https://rdar.apple.com/123119816">rdar://123119816</a>

Reviewed by Jean-Yves Avenard.

When MediaPlayer has a VideoReceiverEndpoint, use it to create a FigVideoTarget and add it to
MediaPlayerPrivateMediaSourceAVFObjC&apos;s AVSampleBufferVideoRenderer. Since
AVSampleBufferDisplayLayer&apos;s video renderer does not support adding video targets, this patch
teaches MediaPlayerPrivateMediaSourceAVFObjC to create an AVSampleBufferVideoRenderer when a
VideoReceiverEndpoint is available. This required abstracting the common interface between
AVSampleBufferDisplayLayer and AVSampleBufferVideoRenderer into @protocol WebSampleBufferVideoRendering
and passing that type to clients of MediaPlayerPrivateMediaSourceAVFObjC.

Note that MediaPlayerPrivateWebM has not been taught to create a AVSampleBufferVideoRenderer;
that will occur in a follow-up fix.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
Declared and soft-linked AVSampleBufferVideoRenderer notification names.

* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.h:
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
Changed to work in terms of WebSampleBufferVideoRendering.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
Taught to ensure and destroy a AVSampleBufferVideoRenderer and changed to work in terms of
WebSampleBufferVideoRendering.

* Source/WebCore/platform/graphics/cocoa/WebSampleBufferVideoRendering.h: Added. Defines a protocol
that encapsulates the common interface subset of AVSampleBufferDisplayLayer and AVSampleBufferVideoRenderer.

Canonical link: <a href="https://commits.webkit.org/274931@main">https://commits.webkit.org/274931@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04a3babc00e4b5381b6cbd4a1e8d747bdc9686d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19425 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42960 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42720 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16754 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33524 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16380 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34842 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14106 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14172 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35811 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44236 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33866 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36641 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36134 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39875 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40039 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15228 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12472 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38157 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16847 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47049 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/9065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16896 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9683 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5356 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16491 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -836,12 +851,22 @@ void getSupportedTypes(HashSet<String>& types) const final
+        return;
+    }
+
     if (shouldEnsureLayer()) {
```
I would tend to merge this into a single method shouldEnsureVideoRendererOrLayer so that we only have to deal between either decompression or the displaying mode.

Otherwise it is starting to look a bit convoluted.

But this can be done as a follow-up/fix

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -971,14 +996,69 @@ void getSupportedTypes(HashSet<String>& types) const final
+    CMTime currentTime = PAL::CMTimebaseGetTime([m_synchronizer timebase]);
+    [m_synchronizer removeRenderer:m_sampleBufferVideoRenderer.get() atTime:currentTime completionHandler:nil];
+
+    if (m_mediaSourcePrivate)
```
if we had combined AVSBDL and AVSBVR in one check, all this is duplicated code.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1597,6 +1677,32 @@ void getSupportedTypes(HashSet<String>& types) const final
 
+WebSampleBufferVideoRendering *MediaPlayerPrivateMediaSourceAVFObjC::sampleBufferVideoRenderer() const
+{
+    if (m_sampleBufferVideoRenderer)
```
one liner?

jyavenard reviewed (approved):
some nits ,but can be done as a followup/bug fix.

In particular this indicates that webm playback is broken on the platform where AVSampleBufferVideoRendering is used.

I would add in the commit description the status of webm and that it will come in a follow-up fix.

aestes committed:
[MediaSource] Use an AVSampleBufferVideoRenderer when a video receiver endpoint is available
https://bugs.webkit.org/show_bug.cgi?id=269621
rdar://123119816

Reviewed by Jean-Yves Avenard.

When MediaPlayer has a VideoReceiverEndpoint, use it to create a FigVideoTarget and add it to
MediaPlayerPrivateMediaSourceAVFObjC's AVSampleBufferVideoRenderer. Since
AVSampleBufferDisplayLayer's video renderer does not support adding video targets, this patch
teaches MediaPlayerPrivateMediaSourceAVFObjC to create an AVSampleBufferVideoRenderer when a
VideoReceiverEndpoint is available. This required abstracting the common interface between
AVSampleBufferDisplayLayer and AVSampleBufferVideoRenderer into @protocol WebSampleBufferVideoRendering
and passing that type to clients of MediaPlayerPrivateMediaSourceAVFObjC.

Note that MediaPlayerPrivateWebM has not been taught to create a AVSampleBufferVideoRenderer;
that will occur in a follow-up fix.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
Declared and soft-linked AVSampleBufferVideoRenderer notification names.

* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.h:
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
Changed to work in terms of WebSampleBufferVideoRendering.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
Taught to ensure and destroy a AVSampleBufferVideoRenderer and changed to work in terms of
WebSampleBufferVideoRendering.

* Source/WebCore/platform/graphics/cocoa/WebSampleBufferVideoRendering.h: Added. Defines a protocol
that encapsulates the common interface subset of AVSampleBufferDisplayLayer and AVSampleBufferVideoRenderer.

Canonical link: https://commits.webkit.org/274931@main
